### DISCUSSION TITLE
Seeking Guidance on Custom Plugin Development

### DISCUSSION THREAD

**Discussion by willieharthga** [ghost-ng/slinger#2] [Q&A]
I hope this message finds you well. I've been following your work on the slinger project. First, let me express my sincere gratitude for maintaining such an excellent project!
I've been using slinger and have some questions about custom plugins. I don't really know how to write and integrate custom plugins properly, and I wonder if you could provide some guidance or insights for me. Are there any specific and more detailed documentation resources I could refer to?
Additionally, I have some ideas that might help to improve the project. When you're available, I'd be more than happy to discuss these ideas with you.
Looking forward to hearing from you.

**Comment by ghost-ng** [comment]
@iamSiddhartha thanks for using slinger!  Did you see the Contributing section in the readme and did you look at the [example plugin](https://github.com/ghost-ng/slinger/blob/main/src/slingerpkg/plugins/my_plugin.py)?  What parts are you having trouble with?

For organization, open up a separate thread for your ideas.

**Reply by willieharthga** [reply]
Thank you very much for your kind reply. The specific situation I encountered is that I wrote the plugin and placed it in the designated directory, but when I actually ran the program, I found that the plugin was not loaded properly. The related plugin did not show up in the command list.

**Reply by ghost-ng** [reply]
@iamSiddhartha  Can you post the code here?  Did you take the example and work off of it?  I recommend first making a direct copy of the example and putting a copy in the `~/.slinger/plugin/` folder - if that doesn't work, then there is another problem.  You can also look at the logs folder in the .slinger folder to try a debug what is going on.

**Reply by willieharthga** [reply]
I attempted to debug using debug mode and also examined the corresponding log files, but I still couldn't determine the cause of the issue. Below is the specific code content of the plugin:

```python
# system_audit.py
from slingerpkg.lib.plugin_base import PluginBase
import argparse
from slingerpkg.utils.printlib import *
import json

class SystemAudit(PluginBase):

    def get_parser(self):
        parser = argparse.ArgumentParser(add_help=False)
        subparsers = parser.add_subparsers(dest='command')
        audit_parser = subparsers.add_parser("audit", help="Perform system audit")
        audit_parser.add_argument("--output", choices=['console', 'json'], default='console', help="Output format")
        audit_parser.set_defaults(func=self.run)
        return parser

    def run(self, args):
        print_info("Starting System Audit...")
        
        audit_data = {
            "system_info": self.get_system_info(),
            "user_accounts": self.get_user_accounts(),
            "installed_software": self.get_installed_software()
        }

        if args.output == 'json':
            print(json.dumps(audit_data, indent=2))
        else:
            self.print_audit_results(audit_data)

    def get_system_info(self):
        print_info("Collecting System Information...")
        self.client.setup_dce_transport()
        self.client.dce_transport._connect('srvsvc')
        response = self.client.dce_transport._enum_info()
        info = response['InfoStruct']['ServerInfo101']
        return {
            "name": info['sv101_name'],
            "platform_id": info['sv101_platform_id'],
            "version": f"{info['sv101_version_major']}.{info['sv101_version_minor']}",
            "type": info['sv101_type'],
            "comment": info['sv101_comment']
        }

    def get_user_accounts(self):
        print_info("Enumerating User Accounts...")
        users = []
        try:
            for user in self.client.enum_users():
                users.append({
                    "name": user['Name'],
                    "full_name": user['FullName'],
                    "comment": user['UserComment'],
                    "flags": user['UserFlags']
                })
        except Exception as e:
            print_error(f"Error enumerating users: {str(e)}")
        return users

    def get_installed_software(self):
        print_info("Listing Installed Software...")
        software = []
        try:
            reg_handle = self.client.reg_open_key(r"HKLM\SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall")
            for subkey in self.client.reg_enum_key(reg_handle):
                try:
                    subkey_handle = self.client.reg_open_key(fr"HKLM\SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\{subkey}")
                    display_name = self.client.reg_query_value(subkey_handle, "DisplayName")
                    if display_name:
                        software.append(display_name)
                except Exception:
                    pass
        except Exception as e:
            print_error(f"Error listing software: {str(e)}")
        return software

    def print_audit_results(self, audit_data):
        print_info("System Information:")
        for key, value in audit_data["system_info"].items():
            print_log(f"{key.capitalize()}: {value}")

        print_info("\nUser Accounts:")
        for user in audit_data["user_accounts"]:
            print_log(f"Name: {user['name']}")
            print_log(f"Full Name: {user['full_name']}")
            print_log(f"Comment: {user['comment']}")
            print_log(f"Flags: {user['flags']}")
            print_log("---")

        print_info("\nInstalled Software:")
        for software in audit_data["installed_software"]:
            print_log(software)
```

**Reply by ghost-ng** [reply]
![image](https://github.com/user-attachments/assets/953e05e1-9aeb-4790-8d97-ab080b5a1372)
![image](https://github.com/user-attachments/assets/ed2fd43d-5034-4355-8360-b1f646cdb175)
![image](https://github.com/user-attachments/assets/da2b71fa-8066-4be5-856e-bc0567e06b9f)

FYI I updated print_error to print_bad.  Also you'll need look at the winreg library in slinger and reproduce how it queries keys.
@iamSiddhartha

**Reply by ghost-ng** [reply]
![image](https://github.com/user-attachments/assets/f70c580f-f756-484f-8f19-36145bcbe3a2)

```
subkeys = self.client.enum_subkeys(r"HKLM\SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall")
print(subkeys)
```

**Reply by ghost-ng** [reply]
if you find a core function that you can add, feel free to make a pull request!  I understand the documentation is lacking.....often that LAST thing we think of...

**Reply by willieharthga** [reply]
Thank you so much for your help! I finally discovered the issue. I had confused the ~/.slinger/plugin/ directory with the directory I cloned from git...... Now I can load the plugin normally. 
Your guidance has been incredibly helpful to me. I have some other ideas and questions about the development process, and I was wondering if you'd be willing to exchange contact information? This way, if I encounter similar issues in the future, I could more easily seek your advice. Perhaps we could also have more in-depth discussions on technical topics.
If you're comfortable with it, maybe we could exchange email addresses or any other contact method you prefer?
Thank you once again for your patient guidance and help!

**Reply by ghost-ng** [reply]
sure!  you can find me on gitter - [@ghost-ng:gitter.im](https://matrix.to/#/@ghost-ng:gitter.im)

**Reply by willieharthga** [reply]
Thank you! Looking forward to our future discussions! Have a great day!Thank you! Looking forward to our future discussions! Have a great day!

**Reply by ghost-ng** [reply]
@iamSiddhartha I released a new version with a plugin inspired by you.  I also annotated requirements in there, hopefully this helps slightly more.
